rustls/src/crypto/aws_lc_rs/ticketer.rs

```diff
@@ -0,0 +1,354 @@
+        let dec_ctx = aes_key
+            .encrypt(&mut encrypted_state)
+            .ok()?;
+        let iv: &[u8] = (&dec_ctx).try_into().ok()?;
+
+        // Produce the MAC tag over the relevant context & encrypted state.
+        // Quoting RFC 5077:
+        //   "The Message Authentication Code (MAC) is calculated using HMAC-SHA-256 over
+        //    key_name (16 octets) and IV (16 octets), followed by the length of
+        //    the encrypted_state field (2 octets) and its contents (variable
+        //    length)."
+        let mut hmac_data =
+            Vec::with_capacity(self.key_name.len() + iv.len() + 2 + encrypted_state.len());
+        hmac_data.extend(&self.key_name);
+        hmac_data.extend(iv);
+        hmac_data.extend(
+            u16::try_from(encrypted_state.len())
+                .ok()?
+                .to_be_bytes(),
+        );
+        hmac_data.extend(&encrypted_state);
+        let tag = hmac::sign(&self.hmac_key, &hmac_data);
+        let tag = tag.as_ref();
+
+        // Combine the context, the encrypted state, and the tag to produce the final ciphertext.
+        // Ciphertext structure is:
+        //   key_name: [u8; 16]
+        //   iv: [u8; 16]
+        //   encrypted_state: [u8, _]
+        //   mac tag: [u8; 32]
+        let mut ciphertext =
+            Vec::with_capacity(self.key_name.len() + iv.len() + encrypted_state.len() + tag.len());
+        ciphertext.extend(self.key_name);
+        ciphertext.extend(iv);
+        ciphertext.extend(encrypted_state);
+        ciphertext.extend(tag);
+
+        self.maximum_ciphertext_len
+            .fetch_max(ciphertext.len(), Ordering::SeqCst);
+
+        Some(ciphertext)
+    }
+
+    fn decrypt(&self, ciphertext: &[u8]) -> Option<Vec<u8>> {
+        if ciphertext.len()
+            > self
+                .maximum_ciphertext_len
+                .load(Ordering::SeqCst)
+        {
+            #[cfg(debug_assertions)]
+            debug!("rejected over-length ticket");
+            return None;
+        }
+
+        // Split off the key name from the remaining ciphertext.
+        let (alleged_key_name, ciphertext) = try_split_at(ciphertext, self.key_name.len())?;
+
+        // checking the key_name is the expected one, *and* then putting it into the
+        // additionally authenticated data is duplicative.  this check quickly rejects
+        // tickets for a different ticketer (see `TicketSwitcher`), while including it
+        // in the AAD ensures it is authenticated independent of that check and that
+        // any attempted attack on the integrity such as [^1] must happen for each
+        // `key_label`, not over a population of potential keys.  this approach
+        // is overall similar to [^2].
+        //
+        // [^1]: https://eprint.iacr.org/2020/1491.pdf
+        // [^2]: "Authenticated Encryption with Key Identification", fig 6
+        //       <https://eprint.iacr.org/2022/1680.pdf>
```

Comment:
I would remove this block as this construction is key-committing.

Reply:
Done :+1: